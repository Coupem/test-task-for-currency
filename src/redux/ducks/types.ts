import {
  LOAD_CURRENCY,
  LOAD_CURRENCY_FAILED,
  LOAD_CURRENCY_SUCCESS,
} from './actionTypes';

export interface ResponseCurrency {
  ccy: string;
  base_ccy: string;
  buy: string;
  sale: string;
}

export interface Currency {
  ccy: string;
  baseCcy: string;
  buy: string;
  sale: string;
}

export interface CurrencyState {
  currency: Currency[];
  isLoadingCurrency: boolean;
  error: string | null;
}

export interface FetchCurrencyFailedPayload {
  error: string;
}

export interface FetchCurrencyRequest {
  type: typeof LOAD_CURRENCY;
}

export type FetchCurrencySuccess = {
  type: typeof LOAD_CURRENCY_SUCCESS;
  payload: Currency[];
};

export type FetchCurrencyFailed = {
  type: typeof LOAD_CURRENCY_FAILED;
  payload: FetchCurrencyFailedPayload;
};

export type CurrencyActions =
  | FetchCurrencyRequest
  | FetchCurrencySuccess
  | FetchCurrencyFailed;

export interface StoreState {
  currency: CurrencyState;
}

// export interface PropSelector {
//   listCurrency: Currency[];
//   baseCurrency: string;
//   inputValue: string;
// }
