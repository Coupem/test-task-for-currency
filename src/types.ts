import { ChangeEventHandler } from 'react';

export const LOAD_CURRENCY_SUCCESS =
  'task-currency/currency/LOAD_CURRENCY_SUCCESS';

export const LOAD_CURRENCY = 'task-currency/currency/LOAD_CURRENCY';

export const LOAD_CURRENCY_FAILED =
  'task-currency/currency/LOAD_CURRENCY_FAILED';

export interface ResponseCurrency {
  ccy: string;
  // eslint-disable-next-line camelcase
  base_ccy: string;
  buy: string;
  sale: string;
}

export interface Currency {
  ccy: string;
  baseCcy: string;
  buy: string;
  sale: string;
}

export interface CurrencyState {
  currency: Currency[];
  isLoadingCurrency: boolean;
  error: string | null;
}

export interface RootState {
  currency: CurrencyState;
}

export interface FetchCurrencyFailedPayload {
  error: string;
}

export interface FetchCurrencyRequest {
  type: typeof LOAD_CURRENCY;
}

export type FetchCurrencySuccess = {
  type: typeof LOAD_CURRENCY_SUCCESS;
  payload: Currency[];
};

export type FetchCurrencyFailed = {
  type: typeof LOAD_CURRENCY_FAILED;
  payload: FetchCurrencyFailedPayload;
};

export type CurrencyActions =
  | FetchCurrencyRequest
  | FetchCurrencySuccess
  | FetchCurrencyFailed;

export interface StoreState {
  currency: CurrencyState;
}

export interface FavoriteState {
  ccy: string;
  buy: string;
  sale: string;
}

interface FavoriteStateForProp {
  ccy?: string;
  buy?: string;
  sale?: string;
}

export interface CurrencyStorage {
  [key: string]: FavoriteState;
}

export interface FavoriteCurrencyProp {
  [key: string]: FavoriteStateForProp;
}

export interface PropConverter {
  baseCurrency: string;
  convertedValue: number;
}

export interface PropsInput {
  baseCurrency: string;
  handleChangeCurrency: ChangeEventHandler<HTMLSelectElement>;
  handleChangeInput: (value: string) => void;
}

export type ErrorsHandler = {
  amount?: string;
};

export interface Primary {
  readonly primary: boolean;
}
