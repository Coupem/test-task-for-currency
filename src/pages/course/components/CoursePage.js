import React, { useState } from 'react';
import { Table, Badge } from 'react-bootstrap';
import { ValidCourseContainer } from './styledComponents';

function CoursePage({ loadedCurrency }) {
  const [bestCurrency, setBestCurrency] = useState([]);
  const filteredCurrency = loadedCurrency.filter((item) => item.ccy !== 'UAH');

  const handleClickSave = (selectedCurrency) => {
    setBestCurrency([{ ...selectedCurrency }]);
  };

  return (
    <ValidCourseContainer>
      <Table
        className="mt-5 w-75 m-auto"
        striped
        bordered
        hover
        variant="primary"
      >
        <thead>
          <tr>
            <th>Currency</th>
            <th>Buy</th>
            <th>Sale</th>
          </tr>
        </thead>
        <tbody>
          {filteredCurrency.map((item) => (
            <tr key={item.ccy}>
              <td>{item.ccy}</td>
              <td>{item.buy}</td>
              <td>
                {item.sale}
                <Badge onClick={() => handleClickSave(item)} variant="dark">
                  save
                </Badge>
              </td>
            </tr>
          ))}
        </tbody>
      </Table>
    </ValidCourseContainer>
  );
}

export default CoursePage;
