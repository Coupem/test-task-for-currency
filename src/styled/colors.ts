export const colorWhite = 'white';

export const colorBlue = '#086d9f';

export const colorRed = '#dd3646';
