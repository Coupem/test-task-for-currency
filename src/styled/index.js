export * from './MainLayout/MainContentWrapper';
export * from './MainLayout/InteractionContainerWrapper';
export * from './Footer/FooterComponent';
export * from './Header/HeaderComponent';
export * from './Navigation/NavButton';
export * from './Navigation/NavigationComponent';
export * from './Course/FavoriteCurrencyWrapper';
export * from './Course/ButtonBest';
export * from './Course/NoFavoriteCurrency';
export * from './Course/TitleCourseComponent';
export * from './Course/TitleFavoriteCurrency';
export * from './Course/ValidCourseContainer';
export * from './Converter/ConverterContainer';
export * from './Converter/TitleConverterContainer';
export * from './Course/ListTableCurrency';
export * from './MainLayout/MainLayoutContainer';
export * from './Converter/InputGroupContainer';
export * from './Converter/TableCurrencyContainer';
export * from './Course/ListGroupContainer';
export * from './Converter/InputContainer';
export * from './Converter/SelectContainer';
export * from './Converter/ErrorContainer';
export * from './Converter/OptionContainer';
