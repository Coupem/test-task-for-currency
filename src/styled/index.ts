export * from './Footer/FooterComponent';
export * from './Header/HeaderComponent';
export { NavButton, NavigationComponent } from './Navigation/navigationStyled';
export {
  MainContentWrapper,
  InteractionContainerWrapper,
  MainLayoutContainer,
} from './MainLayout/mainLayout';
export {
  FavoriteCurrencyWrapper,
  ButtonBest,
  NoFavoriteCurrency,
  TitleCourseComponent,
  TitleFavoriteCurrency,
  ValidCourseContainer,
  ListGroupContainer,
  ListTableCurrency,
} from './Course/courseStyled';
export {
  TitleConverterContainer,
  InputGroupContainer,
  TableCurrencyContainer,
  SelectContainer,
  InputContainer,
  ErrorContainer,
  ConverterContainer,
  OptionContainer,
} from './Converter/converterStyled';
