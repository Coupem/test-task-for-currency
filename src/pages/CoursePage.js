import React, { useEffect, useState } from 'react';
import { useSelector } from 'react-redux';

import {
  getLocalStorageItem,
  setLocalStorageItem,
} from '../utils/loacalStorage';
import { favoriteCurrenciesKey } from '../constants';
import {
  ValidCourseContainer,
  TitleCourseComponent,
  ButtonBest,
  ListTableCurrency,
} from '../styled';
import { FavoriteCurrency } from '../components';

const CoursePage = () => {
  const [favoriteCurrencies, setFavoriteCurrencies] = useState({});

  const { isLoadingCurrency: isLoading, currency: loadedCurrency } =
    useSelector(({ responseData }) => responseData);

  const handleClickSave = ({ ccy, buy, sale }) => {
    if (favoriteCurrencies[ccy]) {
      return;
    }
    const updatedCurrencies = {
      ...favoriteCurrencies,
      [ccy]: { ccy, buy, sale },
    };

    setFavoriteCurrencies(updatedCurrencies);

    setLocalStorageItem(favoriteCurrenciesKey, updatedCurrencies);
  };

  const filteredCurrency = loadedCurrency.filter((item) => item.ccy !== 'UAH');

  useEffect(() => {
    const currencies = getLocalStorageItem(favoriteCurrenciesKey);

    setFavoriteCurrencies(currencies || {});
  }, []);

  if (isLoading) {
    return null;
  }

  return (
    <ValidCourseContainer>
      <TitleCourseComponent>Valid Course:</TitleCourseComponent>
      <FavoriteCurrency favoriteCurrencies={favoriteCurrencies} />
      <ListTableCurrency striped bordered hover className="p-lg-3 ">
        <thead>
          <tr>
            <th>Currency</th>
            <th>Buy</th>
            <th>Sale</th>
          </tr>
        </thead>
        <tbody>
          {filteredCurrency.map((itemCurrency) => (
            <tr key={itemCurrency.ccy}>
              <td className="col-3 w-25">{itemCurrency.ccy}</td>
              <td>{itemCurrency.buy}</td>
              <td>
                {itemCurrency.sale}
                <ButtonBest
                  onClick={() => handleClickSave(itemCurrency)}
                  variant="dark"
                >
                  best
                </ButtonBest>
              </td>
            </tr>
          ))}
        </tbody>
      </ListTableCurrency>
    </ValidCourseContainer>
  );
};

CoursePage.propTypes = {};

export default CoursePage;
