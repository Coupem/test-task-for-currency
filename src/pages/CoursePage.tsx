// eslint-disable-next-line @typescript-eslint/no-use-before-define
import React, { useEffect, useState } from 'react';
import { useSelector } from 'react-redux';

import { FavoriteCurrency } from '../components';
import {
  getLocalStorageItem,
  setLocalStorageItem,
} from '../utils/loacalStorage';
import { favoriteCurrenciesKey, defaultCurrency } from '../constants';
import {
  ValidCourseContainer,
  TitleCourseComponent,
  ButtonBest,
  ListTableCurrency,
} from '../styled';
import { Currency, RootState, FavoriteState } from '../types';

const CoursePage = (): JSX.Element | null => {
  const [favoriteCurrencies, setFavoriteCurrencies] = useState({});

  const { isLoadingCurrency: isLoading, currency: loadedCurrency } =
    useSelector((state: RootState) => state.currency);

  const handleClickSave = ({ ccy, buy, sale }: FavoriteState) => {
    if (Object.prototype.hasOwnProperty.call(favoriteCurrencies, ccy)) {
      return;
    }

    const updatedCurrencies = {
      ...favoriteCurrencies,
      [ccy]: { ccy, buy, sale },
    };

    setFavoriteCurrencies(updatedCurrencies);

    setLocalStorageItem(favoriteCurrenciesKey, updatedCurrencies);
  };

  const filteredCurrency = loadedCurrency.filter(
    (item: Currency) => item.ccy !== defaultCurrency
  );

  useEffect(() => {
    const currencies = getLocalStorageItem(favoriteCurrenciesKey);

    setFavoriteCurrencies(currencies || {});
  }, []);

  if (isLoading) {
    return null;
  }

  return (
    <ValidCourseContainer>
      <TitleCourseComponent>Valid Course:</TitleCourseComponent>
      <FavoriteCurrency favoriteCurrencies={favoriteCurrencies} />
      <ListTableCurrency striped bordered hover className="p-lg-3 ">
        <thead>
          <tr>
            <th>Currency</th>
            <th>Buy</th>
            <th>Sale</th>
          </tr>
        </thead>
        <tbody>
          {filteredCurrency.map((itemCurrency) => (
            <tr key={itemCurrency.ccy}>
              <td className="col-3 w-25">{itemCurrency.ccy}</td>
              <td>{itemCurrency.buy}</td>
              <td>
                {itemCurrency.sale}
                <ButtonBest onClick={() => handleClickSave(itemCurrency)}>
                  best
                </ButtonBest>
              </td>
            </tr>
          ))}
        </tbody>
      </ListTableCurrency>
    </ValidCourseContainer>
  );
};

CoursePage.propTypes = {};

export default CoursePage;
