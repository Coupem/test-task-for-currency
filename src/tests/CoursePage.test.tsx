import { mount, ReactWrapper, shallow } from 'enzyme';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import React from 'react';
import toJson from 'enzyme-to-json';
import CoursePage from '../pages/CoursePage';
import { RootState } from '../types';
import { NavButton } from '../styled';

describe('Course Page', () => {
  let state: RootState;
  let setUp;
  let component: ReactWrapper;

  beforeAll(() => {
    state = {
      currency: {
        currency: [
          { ccy: 'USD', baseCcy: 'UAH', buy: '27.05000', sale: '27.45000' },
          { ccy: 'EUR', baseCcy: 'UAH', buy: '31.85000', sale: '32.45000' },
          { ccy: 'RUR', baseCcy: 'UAH', buy: '0.35000', sale: '0.38000' },
          { ccy: 'BTC', baseCcy: 'USD', buy: '31125.3298', sale: '34401.6803' },
        ],
        isLoadingCurrency: false,
        error: 'string',
      },
    };
    const mockStore = configureStore();

    setUp = (store: RootState) =>
      mount(
        <Provider store={mockStore(store)}>
          <CoursePage {...state} />
        </Provider>
      );

    component = setUp(state);
  });

  it('should render correctly', () => {
    expect(toJson(component)).toMatchSnapshot();
  });

  it('has correct iteration ', () => {
    expect(component.find('tr')).toHaveLength(5);
  });

  it('test button click', () => {
    const mockCallBack = jest.fn();
    const button = shallow(<NavButton onClick={mockCallBack} />);
    button.find('button').simulate('click');
    expect(mockCallBack.mock.calls.length).toEqual(1);
  });

  it('has correct name currency', () =>
    expect(
      component.find('tbody').find('tr').at(0).find('td').at(0).text()
    ).toEqual('USD'));

  it('has correct name currency', () =>
    expect(
      component.find('tbody').find('tr').at(1).find('td').at(0).text()
    ).toEqual('EUR'));

  it('has correct name currency', () =>
    expect(
      component.find('tbody').find('tr').at(2).find('td').at(0).text()
    ).toEqual('RUR'));

  it('has correct name currency', () =>
    expect(
      component.find('tbody').find('tr').at(3).find('td').at(0).text()
    ).toEqual('BTC'));
});
