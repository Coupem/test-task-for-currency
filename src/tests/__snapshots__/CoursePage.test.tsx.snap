// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Course Page should render correctly 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <CoursePage
    currency={
      Object {
        "currency": Array [
          Object {
            "baseCcy": "UAH",
            "buy": "27.05000",
            "ccy": "USD",
            "sale": "27.45000",
          },
          Object {
            "baseCcy": "UAH",
            "buy": "31.85000",
            "ccy": "EUR",
            "sale": "32.45000",
          },
          Object {
            "baseCcy": "UAH",
            "buy": "0.35000",
            "ccy": "RUR",
            "sale": "0.38000",
          },
          Object {
            "baseCcy": "USD",
            "buy": "31125.3298",
            "ccy": "BTC",
            "sale": "34401.6803",
          },
        ],
        "error": "string",
        "isLoadingCurrency": false,
      }
    }
  >
    <styled.div>
      <div
        className="sc-dIsUp hQxPSx"
      >
        <styled.div>
          <div
            className="sc-iqAclL ljtKDH"
          >
            Valid Course:
          </div>
        </styled.div>
        <FavoriteCurrency
          favoriteCurrencies={Object {}}
        >
          <styled.div>
            <div
              className="sc-fujyAs iCXjma"
            >
              <styled.div>
                <div
                  className="sc-crzoAE NUbRz"
                >
                  Favorite Currency:
                </div>
              </styled.div>
              <styled.div>
                <div
                  className="sc-kEqXSa ezRayN"
                >
                  NO CURRENCY
                </div>
              </styled.div>
            </div>
          </styled.div>
        </FavoriteCurrency>
        <Styled(Component)
          bordered={true}
          className="p-lg-3 "
          hover={true}
          striped={true}
        >
          <ForwardRef
            bordered={true}
            className="sc-jrsJWt hCIPZl p-lg-3 "
            hover={true}
            striped={true}
          >
            <table
              className="sc-jrsJWt hCIPZl p-lg-3  table table-striped table-bordered table-hover"
            >
              <thead>
                <tr>
                  <th>
                    Currency
                  </th>
                  <th>
                    Buy
                  </th>
                  <th>
                    Sale
                  </th>
                </tr>
              </thead>
              <tbody>
                <tr
                  key="USD"
                >
                  <td
                    className="ccy col-3 w-25"
                  >
                    USD
                  </td>
                  <td>
                    27.05000
                  </td>
                  <td>
                    27.45000
                    <styled.button
                      onClick={[Function]}
                    >
                      <button
                        className="sc-iCoGMd gGPFyA"
                        onClick={[Function]}
                      >
                        best
                      </button>
                    </styled.button>
                  </td>
                </tr>
                <tr
                  key="EUR"
                >
                  <td
                    className="ccy col-3 w-25"
                  >
                    EUR
                  </td>
                  <td>
                    31.85000
                  </td>
                  <td>
                    32.45000
                    <styled.button
                      onClick={[Function]}
                    >
                      <button
                        className="sc-iCoGMd gGPFyA"
                        onClick={[Function]}
                      >
                        best
                      </button>
                    </styled.button>
                  </td>
                </tr>
                <tr
                  key="RUR"
                >
                  <td
                    className="ccy col-3 w-25"
                  >
                    RUR
                  </td>
                  <td>
                    0.35000
                  </td>
                  <td>
                    0.38000
                    <styled.button
                      onClick={[Function]}
                    >
                      <button
                        className="sc-iCoGMd gGPFyA"
                        onClick={[Function]}
                      >
                        best
                      </button>
                    </styled.button>
                  </td>
                </tr>
                <tr
                  key="BTC"
                >
                  <td
                    className="ccy col-3 w-25"
                  >
                    BTC
                  </td>
                  <td>
                    31125.3298
                  </td>
                  <td>
                    34401.6803
                    <styled.button
                      onClick={[Function]}
                    >
                      <button
                        className="sc-iCoGMd gGPFyA"
                        onClick={[Function]}
                      >
                        best
                      </button>
                    </styled.button>
                  </td>
                </tr>
              </tbody>
            </table>
          </ForwardRef>
        </Styled(Component)>
      </div>
    </styled.div>
  </CoursePage>
</Provider>
`;
